lib/collector.js

```diff
@@ -111,15 +111,12 @@ class Collector {
       this.#fulfill = resolve;
       this.#reject = reject;
       if (!this.done) return;
-      if (this.validate) {
-        try {
-          this.validate(this.data);
-        } catch (err) {
-          this.#cause = err;
-        }
-      }
-      if (this.#cause) reject(this.#cause);
-      else resolve(this.data);
+      if (this.#cause) return void reject(this.#cause);
+      if (!this.validate) return void resolve(this.data);
+      Promise.resolve(this.validate(this.data)).then((result) => {
+        if (result) resolve(result);
```

Comment:
Maybe we need to check this differently?

For example:
```
resolve(result ?? this.data);
```
For those cases where the validator can serialize to "0" or to an empty string "''"

Reply:
I can offer this option. It's almost like you described
```javascript
then(onFulfilled, onRejected = null) {
    this.#fulfill = onFulfilled;
    this.#reject = onRejected;
    if (!this.done) return this;
    if (this.#cause) return Promise.reject(this.#cause).catch(onRejected);
    if (!this.validate) return Promise.resolve(this.data).then(onFulfilled);
    const validateResult = this.validate(this.data);
    return validateResult instanceof Promise
      ? validateResult.then(onFulfilled, onRejected)
      : Promise.resolve(validateResult).then((res) => {
          onFulfilled(res ?? this.data);
        }, onRejected);
  }
```

Reply:
or

```javascript
  then(onFulfilled, onRejected = null) {
    this.#fulfill = onFulfilled;
    this.#reject = onRejected;
    if (!this.done) return this;
    if (this.#cause) return Promise.reject(this.#cause).catch(onRejected);
    if (!this.validate) return Promise.resolve(this.data).then(onFulfilled);
    return Promise.resolve(this.validate(this.data)).then((result) => {
      if (result instanceof Promise) result.then(onFulfilled, onRejected);
      else onFulfilled(result ?? this.data);
    }, onRejected);
  }
```

Reply:
then(onFulfilled, onRejected = null) {
    this.#fulfill = onFulfilled;
    this.#reject = onRejected;
    if (!this.done) return this;
    if (this.#cause) return Promise.reject(this.#cause).catch(onRejected);
    if (!this.validate) return Promise.resolve(this.data).then(onFulfilled);
    const validationRes = this.validate(this.data);
    if (validationRes && validationRes.then) {
      return validationRes.then(onFulfilled, onRejected);
    }
    return Promise.resolve(validationRes).then((res) => {
      onFulfilled(res ?? this.data);
    });
  }

Reply:
```javascript
  then(onFulfilled, onRejected = null) {
    this.#fulfill = onFulfilled;
    this.#reject = onRejected;
    if (!this.done) return this;
    if (this.#cause) return Promise.reject(this.#cause).catch(onRejected);
    if (!this.validate) return Promise.resolve(this.data).then(onFulfilled);
    const res = this.validate(this.data);
    const promise = res && res.then ? res : Promise.resolve(res ?? this.data);
    return promise.then(onFulfilled, onRejected);
  }
```